spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AppConfigurationRefresh.java

```diff
@@ -79,103 +63,94 @@ public void setApplicationEventPublisher(ApplicationEventPublisher applicationEv
      * @return Refresh event was triggered. No other sources need to be checked.
      */
-    private boolean refresh(ConfigStore store, String storeSuffix, String watchedKeyNames) {
-        String storeNameWithSuffix = store.getEndpoint() + storeSuffix;
-        SettingSelector settingSelector = new SettingSelector().setKeyFilter(watchedKeyNames)
-                .setLabelFilter("*");
-
-        ConfigurationSetting revision = clientStore.getRevison(settingSelector, store.getEndpoint());
-
-        String etag = null;
-        // If there is no result, etag will be considered empty.
-        // A refresh will trigger once the selector returns a value.
-        if (revision != null) {
-            etag = revision.getETag();
-        }
+    private boolean refresh(ConfigStore store) {
+        State state = StateHolder.getState(store.getEndpoint());
+
+        Date date = new Date();
+        if (date.after(state.getNotCachedTime())) {
+            for (ConfigurationSetting watchKey : state.getWatchKeys()) {
+                SettingSelector settingSelector = new SettingSelector().setKeyFilter(watchKey.getKey())
+                        .setLabelFilter(watchKey.getLabel());
+
+                ConfigurationSetting revision = clientStore.getRevison(settingSelector, store.getEndpoint());
+
+                String etag = null;
+                // If there is no result, etag will be considered empty.
+                // A refresh will trigger once the selector returns a value.
+                if (revision != null) {
+                    etag = revision.getETag();
+                }
 
-        if (StateHolder.getEtagState(storeNameWithSuffix) == null) {
-            // On startup there was no Configurations, but now there is.
-            if (etag != null) {
-                LOGGER.info("The store " + store.getEndpoint() + " had no keys on startup, but now has keys to load.");
-                return true;
-            }
-            return false;
-        }
+                LOGGER.error(etag + " - " + watchKey.getETag());
+                if (etag != null && !etag.equals(watchKey.getETag())) {
+                    LOGGER.trace(
+                            "Some keys in store [{}] matching the key [{}] and label [{}] is updated, " +
+                                    "will send refresh event.",
+                            store.getEndpoint(), watchKey.getKey(), watchKey.getLabel());
 
-        if (etag != null && !etag.equals(StateHolder.getEtagState(storeNameWithSuffix).getETag())) {
-            LOGGER.trace("Some keys in store [{}] matching [{}] is updated, will send refresh event.",
-                    store.getEndpoint(), watchedKeyNames);
-            if (this.eventDataInfo.isEmpty()) {
-                this.eventDataInfo = watchedKeyNames;
-            } else {
-                this.eventDataInfo += ", " + watchedKeyNames;
-            }
+                    this.eventDataInfo = watchKey.toString();
```

Comment:
If multiple watchKeys getting updated, the eventDataInfo will reflect the last one of them?

Reply:
EventDataInfo will be the watch key highest listed for that config store to be changed.